File: ThreeEngine/lights/EnvironmentMap.cpp
#include <GL/glew.h>
#include "EnvironmentMap.h"

struct EnvironmentMapConst
{
	glm::vec4 shCoefficients[9];
	float diffuseThresh;
	float diffuseHigh;
	float diffuseLow;
	float specularThresh;
	float specularHigh;
	float specularLow;
};

EnvironmentMap::EnvironmentMap(): m_constant(sizeof(EnvironmentMapConst), GL_UNIFORM_BUFFER)
{
	memset(shCoefficients, 0, sizeof(glm::vec4) * 9);
}


EnvironmentMap::~EnvironmentMap()
{
	
}

void EnvironmentMap::updateConstant()
{
	EnvironmentMapConst c;
	memcpy(c.shCoefficients, shCoefficients, sizeof(glm::vec4) * 9);
	c.diffuseThresh = diffuse_thresh;
	c.diffuseHigh = diffuse_high;
	c.diffuseLow = diffuse_low;
	c.specularThresh = specular_thresh;
	c.specularHigh = specular_high;
	c.specularLow = specular_low;
	m_constant.upload(&c);
}

File: webgpu/client/engine/lights/EnvironmentMap.js
import {IndirectLight} from "./IndirectLight.js"

export class EnvironmentMap extends IndirectLight
{
    constructor()
    {
        super();
        this.shCoefficients = new Array(9);
        for (let i=0;i<9;i++)
        {
            this.shCoefficients[i] = [0,0,0];
        }
        this.constant = engine_ctx.createBuffer0(176, GPUBufferUsage.UNIFORM|GPUBufferUsage.COPY_DST);  
    }

    updateConstant()
    {
        const uniform = new Float32Array(44);

        for (let i=0;i<9;i++)
        {
            uniform[i*4] = this.shCoefficients[i][0];
            uniform[i*4 + 1] = this.shCoefficients[i][1];
            uniform[i*4 + 2] = this.shCoefficients[i][2];
            uniform[i*4 + 3] = 0.0;
        }

        uniform[37] = this.diffuse_thresh;
        uniform[38] = this.diffuse_high;
        uniform[39] = this.diffuse_low;
        uniform[40] = this.specular_thresh;
        uniform[41] = this.specular_high;
        uniform[42] = this.specular_low;

        engine_ctx.queue.writeBuffer(this.constant, 0, uniform.buffer, uniform.byteOffset, uniform.byteLength);
    }

}

